<html>
	  <head>
         <title>Allermatch: Allermatch </title>
      </head>
      <link rel="stylesheet" type="text/css" href="{{ url_for('static', filename='css/style.css') }}">
      <body>
        <table border='00' width='100%' height='100%' cellpadding='0' cellspacing='10'>
          <tr height =' 1%'>
           <td colspan='2' valign='top'>

            <table border='0' width='100%' cellpadding='0' cellspacing='0'
            background="{{ url_for('static', filename='images/allermatch_filler.png') }}">
            <tr><td valign='top'><a href='http://allermatch.org'>
              <img border='0' src="{{ url_for('static', filename='images/allermatch_logo.png') }}"></a>
               </td></tr><tr><td height='10'>&nbsp;</td></tr>
            </table>
          </td></tr>
          <tr>
             <td valign='top'>
                  <a href='/index.html'>Home</a>&nbsp;&nbsp;&nbsp;<br>
                  <a href='/allermatchsearch/form'>Search</a>&nbsp;&nbsp;&nbsp;<br>
                  <a href='/database.html'>Databases</a>&nbsp;&nbsp;&nbsp;<br>
                  <a href='/publication.html'>Publication</a>&nbsp;&nbsp;&nbsp;<br>
                  <a href='/help.html'>Introduction</a>&nbsp;&nbsp;&nbsp;<br>
                  <a href='/example.html'>Example</a>&nbsp;&nbsp;&nbsp;<br>
                  <a href='/about.html'>About us</a>&nbsp;&nbsp;&nbsp;<br>
                  <a href='/feedback.html'>Feedback</a>&nbsp;&nbsp;&nbsp;<br>
                  <a href='/disclaimer.html'>Disclaimer</a>&nbsp;&nbsp;&nbsp;<br>
                  <a href='/copyright.html'>Copyright</a>&nbsp;&nbsp;&nbsp;<br>
                  <a href='/thanks.html'>Thanks</a>&nbsp;&nbsp;&nbsp;<br>
                  <a href='/references.html'>References</a>&nbsp;&nbsp;&nbsp;<br>
             </td>
          <td valign='top'>
<!----#include virtual="header.html" -->
<h2>Introduction</h2>

In most nations, genetically engineered foods must be assessed for
their safety before market approval is granted. An important issue in
this safety assessment is the potential allergenicity of transgenic
("foreign") proteins that have been introduced into the food by genetic
engineering. In other words, what is the chance that the foreign
protein may cause allergic reactions after consumption of the
genetically engineered food containing this protein?<br><br>


Potential allergenicity is assessed during a step-by-step procedure
described by the guidelines of the FAO/WHO Codex alimentarius
Commission for the safety assessment of foods derived from genetically
engineered plants and micro-organisms [<a href="references.html#ref_1">1</a>]. One
important step in this procedure is to determine, with the aid of
computer programs, whether the primary structure (amino acid sequence)
of the transgenic protein is similar to sequences of allergenic
proteins, of which the latter are available from public protein
sequence databases.<br><br>


Two types of similarity are searched for:<br>

<ul>
  <li>Short identical stretches of 6-8 contiguous amino acids;</li>
  <li>larger stretches (80 amino acids long) containing a minimum
      of 35% (non contiguous) identical amino acids.<br> </li>
</ul>

The similar stretches that are identified this way may harbour
potential binding sites (called epitopes) for IgE antibodies. IgE
antibodies are allergy-related and involved in the binding of the
allergen to mast cells, after which these cells release compounds, such
as histamine, that cause the symptoms of allergy. Allergens must at
least contain two IgE-binding epitopes to trigger a mast cell reaction.<br><br>


To search for the two types of similarities, a recent Expert
Consultation of the FAO/WHO, which was held in preparation of the Codex
alimentarius guidelines, devised the following procedure [<a
href="references.html#ref_2">2</a>]:<br><br>

<p align="lefy" style="margin-left: 40px;"> 
<cite>6.1. Sequence Homology as
Derived from Allergen Databases <br><br>


The commonly used protein databases (PIR, SwissProt and TrEMBL) contain
the amino acid sequences of most allergens for which this information
is known. However, these databases are currently not fully up-to-date.
A specialized allergen database is under construction.<br><br>


Suggested procedure on how to determine the percent amino acid identity
between the expressed protein and known allergens.<br><br>


Step 1: obtain the amino acids sequences of all allergens in the
protein databases (for SwissProt and TrEMBL: see <a
href="http://expasy.ch/tools">http://expasy.ch/tools</a>; for PIR see <a
href="http://wwwnbrf.georgetown.edu/pirwww">http://wwwnbrf.georgetown.edu/pirwww</a>
) in FASTA-format (using the amino acids from the mature proteins only,
disregarding the leader sequences, if any). Let this be data set (1).<br><br>


Step 2: prepare a complete set of 80-amino acid length sequences
derived from the expressed protein (again disregarding the leader
sequence, if any). Let this be data set (2).<br><br>


Step 3: go to EMBL internet address: <a href="http://www2.ebi.ac.uk">http://www2.ebi.ac.uk</a>
and compare each of the sequences of the data set (2) with all
sequences of data set (1), using the FASTA program on the web site for
alignment with the default settings for gap penalty and width.<br><br>


Cross-reactivity between the expressed protein and a known allergen (as
can be found in the protein databases) has to be considered when there
is:<br><br>


 1) more than 35 % identity in the amino acid sequence of the expressed
 protein (i.e. without the leader sequence, if any), using a window of
 80 amino acids and a suitable gap penalty (using Clustal-type alignment
 programs or equivalent alignment programs)<br><br>


or:<br><br>


  2) identity of 6 contiguous amino acids.<br><br>


If any of the identity scores equals or exceeds 35 %, this is
considered to indicate significant homology within the context of this
assessment approach. The use of amino acid sequence homologies to
identify prospective cross-reacting allergens in genetically modified
foods has been discussed in more detail elsewhere (Gendel, 1998a;
Gendel, 1998b).<br></cite></p><br>


The search facility on the Allermatch<sup>tm</sup> webtool
automatically carries out
the procedure recommended by the guidelines on protein sequences that
are entered by the user in FASTA format (one-letter code
without residue numbers, see example sequence below). The user has
the option to select the following outputs of interest:<br><br>


<ol>
    <li>Alignment of 80-amino-acid subsequences of the input
    sequence
    using a sliding window of 80-amino acids size. The step size is 1 amino
    acid, such that from a sequence of 100 amino acids, for example, 21
    subsequences of 80 amino acids length are made (1-80, 2-81, 3-82 ...
    20-99, 21-100). Each of these subsequences is aligned to database
    sequences. The FASTA computer algorithm is used for these sequence
    alignments, as recommended (see above; default values are used). With
    FASTA, "head to tail" alignments (from the start to the end of a
    sequence) are made of each subsequence with each database sequence. The
    default threshold for the number of identical amino acids is 35% in the
    alignment with an 80-amino-acid window, which is considered a
    significant level of identity between the input sequence and the
    allergenic protein's sequence (see recommendations cited above). The
    identity presented by the website in the results of the alignments is
    therefore the % identical amino acids in the 80-amino-acid window. The
    default threshold can be changed by the user. Input sequences shorter
    than 80 amino acids should not be aligned using this option.</li>
    <li>Full alignment of the whole input sequence with database
    sequences using the FASTA algorithm. This option can be used, for
    example, for input sequences shorter than 80 amino acids, for which the
    option of the 80-amino acids sliding window (see above) cannot be used.
    Also in case where an input sequence shows sufficient identity with
    many proteins over its entire sequence, this option may provide for a
    good oversight of the alignments between the input- and database-
    sequences.</li>

    <li>Exact hits of short identical stretches of, for example, 6
    amino
    acids. To this end, a wordmatch algorithm is used, which searches for
    identical matches of a specified number of contiguous amino acids
    ("wordlength") between the input sequence and a given database
    sequence. The default value for the wordlength, which can be changed by
    the user, is 6 amino acids. Decreasing the wordlength likely results in
    a larger number of positive scores, while increasing it may yield less
    positive results.<br></li>

</ol><br>


The entered sequences will be compared to the sequences of allergenic
proteins compiled in the database. These sequences of allergenic
proteins have been extracted from protein databases. 
Putative signal-, pro-, and transit-peptides, have been removed from these sequences, which yields the
sequences of "mature" proteins.

Positive results of the analysis will be provided to the user. 
<br><br>
<!--#include virtual="footer.html" -->
</td></tr>
     <tr><td valign='bottom' colspan='2'>
        <table width='100%'><tr>
           <td valign='bottom'>
           <td valign='bottom'>
               <a href='mailto:Gijs.Kleter@wur.nl'>Questions: Dr. Gijs Kleter</a><br>
           </td>
           <td valign='bottom' align='right'>
            <img align='right' src='/static/images/wageningenur.gif'>
          </td></tr>
     </table>
  </td></tr>
</table>
</body>
</html>
